File: services/dotnet-microservices/ProductService/Application/ProductServiceImpl.cs
namespace ProductService.Application;

using ProductService.Domain.Interfaces;
using ProductService.Domain.Models;
using ProductService.Infrastructure.Cache;
using ProductService.Infrastructure.Messaging;

public class ProductServiceImpl : IProductService
{
    private readonly IProductRepository _repository;
    private readonly IRedisCacheService _cache;
    private readonly KafkaProducer _kafkaProducer;
    private readonly ILogger<ProductServiceImpl> _logger;

    public ProductServiceImpl(
        IProductRepository repository,
        IRedisCacheService cache,
        KafkaProducer kafkaProducer,
        ILogger<ProductServiceImpl> logger)
    {
        _repository = repository;
        _cache = cache;
        _kafkaProducer = kafkaProducer;
        _logger = logger;
    }

    public async Task<Product?> GetProductByIdAsync(Guid id)
    {
        var cacheKey = $"product:{id}";
        var cached = await _cache.GetAsync<Product>(cacheKey);
        if (cached != null) return cached;

        var product = await _repository.GetByIdAsync(id);
        if (product != null)
        {
            await _cache.SetAsync(cacheKey, product, TimeSpan.FromMinutes(10));
        }
        return product;
    }

    public async Task<IEnumerable<Product>> GetAllProductsAsync() => await _repository.GetAllAsync();

    public async Task<IEnumerable<Product>> SearchProductsAsync(string query) => await _repository.SearchAsync(query);

    public async Task<Product> CreateProductAsync(CreateProductRequest request)
    {
        var product = new Product(
            Id: Guid.NewGuid(),
            Name: request.Name,
            Description: request.Description,
            Price: request.Price,
            CategoryId: request.CategoryId,
            ImageUrl: request.ImageUrl,
            IsActive: true,
            CreatedAt: DateTime.UtcNow,
            UpdatedAt: DateTime.UtcNow
        );

        var created = await _repository.CreateAsync(product);

        await _kafkaProducer.PublishAsync("product-events", new
        {
            EventType = "ProductCreated",
            ProductId = created.Id,
            Name = created.Name,
            Price = created.Price,
            Timestamp = DateTime.UtcNow
        });

        _logger.LogInformation("Product created: {ProductId}", created.Id);
        return created;
    }

    public async Task<Product> UpdateProductAsync(Guid id, UpdateProductRequest request)
    {
        var product = await _repository.GetByIdAsync(id)
            ?? throw new KeyNotFoundException($"Product {id} not found");

        var updated = product with
        {
            Name = request.Name ?? product.Name,
            Description = request.Description ?? product.Description,
            Price = request.Price ?? product.Price,
            CategoryId = request.CategoryId ?? product.CategoryId,
            ImageUrl = request.ImageUrl ?? product.ImageUrl,
            IsActive = request.IsActive ?? product.IsActive,
            UpdatedAt = DateTime.UtcNow
        };

        var result = await _repository.UpdateAsync(updated);
        await _cache.RemoveAsync($"product:{id}");

        return result;
    }

    public async Task<bool> DeleteProductAsync(Guid id)
    {
        var deleted = await _repository.DeleteAsync(id);
        if (deleted) await _cache.RemoveAsync($"product:{id}");
        return deleted;
    }

    public async Task<Category> CreateCategoryAsync(CreateCategoryRequest request)
    {
        var category = new Category(Guid.NewGuid(), request.Name, request.Description, request.ParentId, DateTime.UtcNow);
        return await _repository.CreateCategoryAsync(category);
    }

    public async Task<IEnumerable<Category>> GetAllCategoriesAsync() => await _repository.GetAllCategoriesAsync();

    public async Task<Review> AddReviewAsync(Guid productId, CreateReviewRequest request)
    {
        _ = await _repository.GetByIdAsync(productId)
            ?? throw new KeyNotFoundException($"Product {productId} not found");

        var review = new Review(Guid.NewGuid(), productId, request.UserId, request.Rating, request.Comment, DateTime.UtcNow);
        return await _repository.AddReviewAsync(review);
    }

    public async Task<IEnumerable<Review>> GetProductReviewsAsync(Guid productId) =>
        await _repository.GetReviewsByProductAsync(productId);
}

File: services/nodejs-microservices/product-service/src/application/ProductServiceImpl.ts
/**
 * Product Service Implementation
 * 
 * Implements Cache-Aside pattern for product reads:
 * - Read: Check Redis → if miss, query DB → store in Redis
 * - Write: Update DB → invalidate Redis cache
 * 
 * This ensures high read performance while maintaining consistency.
 */
import { v4 as uuidv4 } from 'uuid';
import { Product, Category, ProductReview, ProductStatus } from '../domain/models/Product';
import { IProductService, CreateProductDTO, UpdateProductDTO, CreateCategoryDTO, CreateReviewDTO } from '../domain/interfaces/IProductService';
import { ProductRepository } from '../infrastructure/persistence/ProductRepository';
import { KafkaProducer } from '../infrastructure/messaging/KafkaProducer';
import { RedisClient } from '../infrastructure/cache/RedisClient';
import { PRODUCT_EVENTS_TOPIC, ProductEventType } from '../infrastructure/messaging/events';

export class ProductServiceImpl implements IProductService {
  constructor(
    private readonly productRepository: ProductRepository,
    private readonly kafkaProducer: KafkaProducer,
    private readonly redisClient: RedisClient
  ) {}

  async createProduct(dto: CreateProductDTO): Promise<Product> {
    const product: Omit<Product, 'createdAt' | 'updatedAt'> = {
      id: uuidv4(),
      name: dto.name,
      description: dto.description,
      price: dto.price,
      categoryId: dto.categoryId,
      imageUrl: dto.imageUrl,
      status: ProductStatus.ACTIVE,
    };

    const created = await this.productRepository.create(product);

    // Cache the new product
    await this.redisClient.set(`product:${created.id}`, created);

    await this.kafkaProducer.publish(PRODUCT_EVENTS_TOPIC, {
      type: ProductEventType.PRODUCT_CREATED,
      data: { id: created.id, name: created.name, price: created.price },
    });

    return created;
  }

  async getProductById(id: string): Promise<Product | null> {
    // Cache-Aside: check cache first
    const cached = await this.redisClient.get<Product>(`product:${id}`);
    if (cached) {
      console.log(`[Cache] Hit for product:${id}`);
      return cached;
    }

    // Cache miss: load from database
    const product = await this.productRepository.findById(id);
    if (product) {
      // Store in cache for future requests
      await this.redisClient.set(`product:${id}`, product);
    }
    return product;
  }

  async updateProduct(id: string, dto: UpdateProductDTO): Promise<Product | null> {
    const existing = await this.productRepository.findById(id);
    if (!existing) return null;

    const updated = await this.productRepository.update(id, {
      name: dto.name,
      description: dto.description,
      price: dto.price,
      categoryId: dto.categoryId,
      imageUrl: dto.imageUrl,
      status: dto.status as ProductStatus,
    });

    if (updated) {
      // Invalidate cache on write
      await this.redisClient.delete(`product:${id}`);

      // Publish price change event if price changed
      if (dto.price && dto.price !== existing.price) {
        await this.kafkaProducer.publish(PRODUCT_EVENTS_TOPIC, {
          type: ProductEventType.PRICE_CHANGED,
          data: { id: updated.id, oldPrice: existing.price, newPrice: updated.price },
        });
      }

      await this.kafkaProducer.publish(PRODUCT_EVENTS_TOPIC, {
        type: ProductEventType.PRODUCT_UPDATED,
        data: { id: updated.id, name: updated.name, price: updated.price },
      });
    }

    return updated;
  }

  async deleteProduct(id: string): Promise<boolean> {
    const deleted = await this.productRepository.delete(id);
    if (deleted) {
      await this.redisClient.delete(`product:${id}`);
      await this.kafkaProducer.publish(PRODUCT_EVENTS_TOPIC, {
        type: ProductEventType.PRODUCT_DELETED,
        data: { id },
      });
    }
    return deleted;
  }

  async listProducts(page: number, limit: number): Promise<Product[]> {
    const offset = (page - 1) * limit;
    return this.productRepository.findAll(offset, limit);
  }

  async searchProducts(query: string): Promise<Product[]> {
    return this.productRepository.search(query);
  }

  async createCategory(dto: CreateCategoryDTO): Promise<Category> {
    const category: Omit<Category, 'createdAt'> = {
      id: uuidv4(),
      name: dto.name,
      description: dto.description,
      parentId: dto.parentId,
    };
    return this.productRepository.createCategory(category);
  }

  async listCategories(): Promise<Category[]> {
    return this.productRepository.findAllCategories();
  }

  async addReview(productId: string, dto: CreateReviewDTO): Promise<ProductReview> {
    const product = await this.productRepository.findById(productId);
    if (!product) throw new Error('Product not found');

    const review: Omit<ProductReview, 'createdAt'> = {
      id: uuidv4(),
      productId,
      userId: dto.userId,
      rating: dto.rating,
      comment: dto.comment,
    };

    return this.productRepository.createReview(review);
  }

  async getReviews(productId: string): Promise<ProductReview[]> {
    return this.productRepository.findReviewsByProductId(productId);
  }
}
